### PR TITLE
feat: refactor entire sdk to be async

### PR DESCRIPTION
**Pull request by kneekey23**
CorrespondingPr:
https://github.com/awslabs/smithy-swift/pull/389

## Issue \#

Closes #497 

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AWSClientRuntime/Sources/Middlewares/RetryerMiddleware.swift
Comment (wooj2) [inline AWSClientRuntime/Sources/Middlewares/RetryerMiddleware.swift:10]: if we are going to use this, we should scope it to the function it's being used in.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/RetryMiddleware.kt
Comment (wooj2) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/RetryMiddleware.kt:32]: It's interesting that we now need to pass these in -- is this because we are now throwing specific exceptions in the Retryer's implementation?

File: .github/workflows/continuous-integration.yml
Comment (kneekey23) [reply .github/workflows/continuous-integration.yml:75]: why? we want the other linux builds to also run even if one fails cuz different flavors of linux could perform differently?

File: .github/workflows/continuous-integration.yml
Comment (kneekey23) [reply .github/workflows/continuous-integration.yml:79]: it is running there, I just thought we should also run it here. and now that I have ive actually caught a compiler bug,

File: .github/workflows/continuous-integration.yml
Comment (kneekey23) [reply .github/workflows/continuous-integration.yml:88]: the crt builder made some major updates and since we had to update our version to use the latest swift images, we had to also update our run script. one of the changes they made is they no longer store their images in docker, and now use ECR. this change accounts for that.

File: .github/workflows/continuous-integration.yml
Comment (kneekey23) [reply .github/workflows/continuous-integration.yml:98]: the new changes as mentioned above require credentials now. I did discuss with them making changes to the builder to remove this and we will do so in future iterations as technically they should not be required.

File: AWSClientRuntime/Sources/Middlewares/RetryerMiddleware.swift
Comment (kneekey23) [reply AWSClientRuntime/Sources/Middlewares/RetryerMiddleware.swift:10]: will do

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/RetryMiddleware.kt
Comment (kneekey23) [reply codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/RetryMiddleware.kt:32]: previously error type was passed throughout the middleware, now since it doesn't have to be since everything is async throws the middlewares that throw the specific type of error have to know about it thus we have to pass it in as a generic.

**Comment by kneekey23** [comment]
This PR is blocked by a swift compiler bug on Linux reported here: https://bugs.swift.org/browse/SR-15659
Apple will prioritize this bug in January for us.

File: .github/workflows/continuous-integration.yml
Comment (wooj2) [reply .github/workflows/continuous-integration.yml:75]: I see.  Cool, lets keep it.

File: .github/workflows/continuous-integration.yml
Comment (wooj2) [reply .github/workflows/continuous-integration.yml:97]: Hmm maybe add a #todo here to keep track of what changes are needed in the CRT so that we can eventually delete this?

File: AWSClientRuntime/Sources/Auth/AWSCredentialsProvider.swift
Comment (wooj2) [inline AWSClientRuntime/Sources/Auth/AWSCredentialsProvider.swift:84]: Spacing seems off here.

File: AWSClientRuntime/Sources/Auth/CredentialsProviderCRTAdapter.swift
Comment (wooj2) [inline AWSClientRuntime/Sources/Auth/CredentialsProviderCRTAdapter.swift:24]: nit:
```
return credentials.toCRTType()
```

File: builder.json
Comment (wooj2) [inline builder.json:15]: Is this an optimization?

File: AWSClientRuntime/Sources/Auth/AWSCredentialsProvider.swift
Comment (kneekey23) [reply AWSClientRuntime/Sources/Auth/AWSCredentialsProvider.swift:84]: fixed

File: AWSClientRuntime/Sources/Auth/CredentialsProviderCRTAdapter.swift
Comment (kneekey23) [reply AWSClientRuntime/Sources/Auth/CredentialsProviderCRTAdapter.swift:24]: corrected

File: builder.json
Comment (kneekey23) [reply builder.json:15]: yes it is, no need to actually build it agian since we built it before we can actually just run the tests :)

**Comment by kneekey23** [comment]
Update here for anyone following this PR, apple will do a release end of Jan/early Feb and once that happens, this PR will be merged in and released! https://forums.swift.org/t/development-open-for-swift-5-5-3-for-linux-and-windows/54553

File: AWSClientRuntime/Sources/Middlewares/RetryerMiddleware.swift
Comment (kneekey23) [reply AWSClientRuntime/Sources/Middlewares/RetryerMiddleware.swift]: john resolved with some brilliance. code is committed now

**Review by wooj2 — APPROVED** [review]
lgtm!

**Pull request merged**
